### ISSUE TITLE
New Case Action Top-up

### ISSUE THREAD

**Issue by RasmusBurge-CG** [Clinical-Genomics/cg#3971]
As a member of prod bioinfo,
I want a new action on case called `Top-up`,
So that analysis would start automatically when new data is available for cases that have been topped up.

# Clarification

It’s very easy to overlook manually starting cases after a top-up, which has the potential to cause significant delays in processing samples (potentially devastating for critically ill patients). To prevent this and avoid unnecessary sample processing, one could set the case status to Top-up. This way, when new data becomes available for that case, the automation will pick up the sample and start it. Additionally, there is potential for integration with LIMS: when a sample is requeued for top-up, it could automatically update the status in StatusDB.

# Work impact

Answer the following questions:
- Is there currently a workaround for this issue? If so, what is it?
    - Yes, we need to communicate with each other in production to inform one another about a top-up. Alternatively, one could check all the samples in the delivery step, retrieve the case from statusdb, and verify if it’s running in TB.
- How much time would be saved by implementing this feature on a weekly basis?
    - 1 hour
- How many users are affected by this issue?
    - All prod members
- Are customers affected by this issue?
    - Yes, it happens from time to time that we forget to manually restart cases that have been topped up.

# Acceptance Criteria

- [ ] A more secure approach ensures that top-ups are started.

# Notes

- Additional information.
- Dependencies.
- Related user stories.

**Comment by diitaz93** [comment]
# Second clarification
There are two situations (that I know) in which a sample will be sent for top-up:

1. The sample didn't reach the target amount of reads (based on the app tag). In this case, the automation checks that the reads are not reached and simply skips the analysis until the sample has reached the desired amount of reads.
2. The sample got the required amount of reads and it starts the analysis, but during the analysis we realise that it didn't get the expected coverage (or didn't satisfy other pipeline-specific requirements related to sequencing quality). Then it is sent to top-up and the sample/analysis is labelled as analysed/complete (unclear how it works on this level).

The problem described in this issue concerns only the second situation, as in the first one, the sample will be taken by the automation when the desired number of reads is reached. In the second case, the automation will never start again the analysis as it is labelled as already run.

**Comment by henrikstranneheim** [comment]
Will the automation not already pick this up whem there is new seq data for a sample which is newer than the latest analysis and action is None?
```python
    def _is_latest_analysis_done_on_all_sequences(self, case: Case) -> bool:
        return case.latest_analyzed < case.latest_sequenced

**Comment by beatrizsavinhas** [comment]
Yes, I was thinking exactly of this @henrikstranneheim!
The only exception I can think of is if the initial analysis fails, and the case was not stored and therefore still has status `running` in status DB. I think this might be the case with samples that received enough reads but the analysis failed on coverage and was left as `failed` in trailblazer.

**Comment by henrikstranneheim** [comment]
But should that case not warrant a manual investigation? It might fail for  number of reasons.

**Comment by RasmusBurge-CG** [comment]
Hi!

Yes, @beatrizsavinhas, this would fall under case 2, as stated in the second clarification by @diitaz93, right?

@henrikstranneheim, that’s true. If the investigation reveals a need for a top-up, then the Top-up action could be useful. Do you agree?

There might be a better solution to this. If there’s something I’m unaware of, I’m happy to learn!

**Comment by henrikstranneheim** [comment]
> Hi!
> 
> Yes, @beatrizsavinhas, this would fall under case 2, as stated in the second clarification by @diitaz93, right?
> 
> @henrikstranneheim, that’s true. If the investigation reveals a need for a top-up, then the Top-up action could be useful. Do you agree?


Not sure, it would not have any function in cg and it would complicate the logic that finds cases to start, (which is already complicated). Unsure of that it is worth it compare to just a comment in TB and let the automation pick it up once ready.

**Comment by RasmusBurge-CG** [comment]
Okay, I see. It’s just that the automation doesn’t pick it up for pipelines like mip-dna. I do agree that it would affect the start logic and require changes in CG. The benefit, as I see it, would be that a sample could be started during the nighttime, reducing turnaround time. Additionally, in my view, it would lessen the risk of missing a case restart.

**Comment by karlnyr** [comment]
It would be vital that we restart cases whenever new data is available. @Karl-Svard brought up that if we re-demultiplex we would not want to start things because of the new "latest_sequenced_at" date. We have refined the acceptance criteria a bit.

**Comment by beatrizsavinhas** [comment]
Suggestion for acceptance criteria:

- There is a manual way to set an analysis to be started when new sequencing data is available.

Considerations:
- If the case was already analysed after passing aggregated sequencing QC (all samples received enough reads), no analyses should be started automatically. Example: sample A received enough reads and is in the same pool as sample B that needs top up. The whole pool is re-sequenced since more reads are needed for sample B. No new analyses for sample A should start.

**Comment by henrikstranneheim** [comment]
Hhmm, tricky one. @RasmusBurge-CG Do you have an example of a case that is not picked up and do you have any reason why.

@beatrizsavinhas Can't you set the case action to "analyze"?

**Comment by beatrizsavinhas** [comment]
I don't think that would work as intended @henrikstranneheim because if the case is set to `analyse` the analysis will just start again when the systemd runs, with no regard to there being new sequencing data or not. So we run the risk of analysing the case again with the same data. What we need is something that makes it so the analysis only starts after the sample was topped up.

But as @karlnyr mentioned, we will still discuss this issue in our next meeting!

**Comment by islean** [comment]
I think setting the case action to `None` will result in the behaviour you are after:

1. ANALYZE will result in it getting picked up for analysis regardless of whether there is new sequencing data or not
2. RUNNING will result in the case never being picked up
3. None will result in a check if there is new sequencing data since the latest analysis and if so start it

**Comment by karlnyr** [comment]
That would restart the case instantly. Since there is no analysis, and there is sequence data. What we want, is to queue the analysis for start after it has been sequenced again. Meaning, that there is not analysis object, there is sequencing data, but what we need is to check if there is new sequencing data that has been added to the sample. This could be done in a lot of ways- perhaps checking if there is an analysis already in trailblazer that is failed - if so, it means that we have once started the case but something was up. Or, you could check if more than one flow cell exists on the sample. But it needs to be a combination.

**Comment by karlnyr** [comment]
@Clinical-Genomics/sysdev We have tried to clarify this user story and updated the acceptance criteria

**Comment by RasmusBurge-CG** [comment]
Additional benefit of having a new action `Top-up` would be that this is also seen in the orderportal. In this way the customer would be aware of the status of the sample and case.

**Comment by diitaz93** [comment]
## Technical Refinement
Findings:
The case is not picked up to analyze because the `action` attribute of the `Case` in cg is kept as `running` and not changed when the analysis in Trailblazer is set to FAILED. The case is filtered out when calling the ReadHandler function `get_cases_to_analyze`.
Implementation ideas:
1. Setting the `Case.action` to None will let the case be picked up for analysis. 
	1. Pros: simple to implement
	2. Cons: Ambiguous, because None is a value used for `Case.action` in different situations, for example during lab, sequencing and after storing.
2. Create a new value in the `CaseActions` StrEnum with a meaningful value for the top-up stage and set cases sent to top up with this value.
	1. Pros: We can identify in which stage of the flow are top-up cases. It would de easier for users to understand what is happening
	2. Cons: More invasive solution, involves an alembic migration and modifying Read handler functions.

In any case, we need to set `Case.action` to another value manually. It is unclear for now if this should be implemented in any WebUI or as a CLI command.

**Comment by diitaz93** [comment]
## Decision after Refinement
We are leaning towards adding an analysis entry to the statusDB when the analysis is starting (sent to TB and its jobs to SLURM). This option would include a column where we state if an analysis has failed or not. More digging into the issue is required to see how viable is this option.

**Comment by mogmi** [comment]
## Refinement round 2

Suggestions:
- Create a migration that: 
  - adds the boolean mandatory field `is_complete` to the Analysis table
  - set `is_complete` to True for all existing Analysis entries
  - introduce a new value for case action: "top-up" (also add it to the enum `CaseAction`)
- Move the creation of Analysis models to `start_available` instead of `store_available` (except for raw-data where it can stay in `store_available` where there is no `start_available`) in here `is_complete` will be set to False. 
- Set `is_complete` = True for the analysis in `store_available`
- Modify the filter in `get_cases_ready_for_analysis`:
  - Add a clause to the `or`-statement so that cases with the action `top-up` gets picked up. 
  - To prevent re-analysis of samples in the same pool as a sample set to top-up: add `Analysis._is_completed(False)` to the third clause. 
- To preserve the current behavior, add `Analysis.is_complete == True` to the existing analysis filters (rename crud-functions if necessary to reflect this).
- The manual marking to set a case for top-up will be to set the `case.action` to `top-up` in status-db
